### ISSUE TITLE
Unable to set particular TLS 1.3 cipher(s)

### ISSUE THREAD

**Issue by Tantrix** [luadch/luadch#102]
>             --// use this config for:  TLSv1.3 with strongest available cipher; requires OpenSSL 1.1.1
>                 protocol = "tlsv1_3",
> 				--ciphers = "HIGH+kEDH:HIGH+kEECDH:HIGH:!PSK:!SRP:!3DES:!aNULL",

There are only 5 (five) TLS 1.3 ciphers.
There are only 3 (three) enabled by default.
There are only 2 (two) that (most) DC clients can use.
There is one best for running DC hub.

It is **TLS_AES_128_GCM_SHA256** (or **ChaCha** with **Poly1305**)

So, what's the reason in using such a long string?

**Comment by Sopor** [comment]
Those are cipher suites and those starting with `!` are disabled if i understand this correct :)

Do you have some issues with the cipher? I can see that you still have comment it out with two dashes but not the protocol.

**Comment by Tantrix** [comment]
> those starting with ! are disabled
 
That's what I'm talking about. No need to disable what TLS 1.3 does not have at all. You can check https://wiki.openssl.org/index.php/TLS1.3#Ciphersuites for details. 

So I'm curious why it is written in that way and how to specify one preferred TLS 1.3 cipher directly. 

> you still have comment it out with two dashes but not the protocol.

Yeah. I tried

```
            --// use this config for:  TLSv1.3 with strongest available cipher; requires OpenSSL 1.1.1
                protocol = "tlsv1_3",
				--ciphers = "HIGH+kEDH:HIGH+kEECDH:HIGH:!PSK:!SRP:!3DES:!aNULL",
                ciphers = "TLS_AES_128_GCM_SHA256",
```
with no success (hub refuses to start).

Something TLS 1.2 related like

```
            --// use this config for:  TLSv1.3 with strongest available cipher; requires OpenSSL 1.1.1
                protocol = "tlsv1_3",
				--ciphers = "HIGH+kEDH:HIGH+kEECDH:HIGH:!PSK:!SRP:!3DES:!aNULL",
                ciphers = "ECDHE+AES128",
```
makes the hub working but not specifies the cipher.

**Comment by Tantrix** [comment]
Oops. Seems like I'm wrong in some cases.

First, TLS 1.3 cipher is something that shouldn't be just set by the hub admin software-side. SSL library itself, hub server CPU, CPU of the connecting client – it's all matters.

Second, there is a fresh attempt to make cipher priority working IRL: https://sourceforge.net/p/dcplusplus/code/ci/454a53433a701bea163cef51bd79696b1fdb2af2/ inspired by the https://klondike.es/klog/2021/04/12/some-notes-on-cipher-choices-and-tls1-3/ article.

Be sure to check both before making any TLS 1.3 ciphers decisions.

**Comment by pulsar-de** [comment]
> First, TLS 1.3 cipher is something that shouldn't be just set by the hub admin software-side. SSL library itself, hub server CPU, CPU of the connecting client – it's all matters.

I think the hub admin should set the rules for tolerated ciphers. what if the client wants to use an insecure cipher, should the hub allow that? I do not think so. the hub decides what's acceptable and what's not and the client adapts. 

It works a bit differently with TLS 1.3. It is entirely possible to define which ciphers are acceptable for the hub and which are not. There are two possibilities:

1. The hub defines a number of tolerated ciphers according to priority:
```lua
            --// use this config for:  TLSv1.3 with AES256 + CHACHA20 + AES128; requires OpenSSL 1.1.1 or higher
                protocol = "tlsv1_3",
                ciphers = "HIGH+kEDH:HIGH+kEECDH:HIGH:!PSK:!SRP:!3DES:!aNULL",
                ciphersuites = "TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256",
```
Important: With this method, the client decides which cipher he wants to use; the priority order set by the hub does not have to be observed. 

2. The hub strictly expects only one specified cipher: 
```lua
            --// use this config for:  TLSv1.3 with forced ciphersuite; requires OpenSSL 1.1.1 or higher
                protocol = "tlsv1_3",
                ciphers = "HIGH+kEDH:HIGH+kEECDH:HIGH:!PSK:!SRP:!3DES:!aNULL",
                ciphersuites = "TLS_AES_256_GCM_SHA384", -- force AES256
```
I will also include that in the next release, but you should note that stronger encryption is always at the expense of performance.

**Comment by Tantrix** [comment]
> With this method, the client decides which cipher he wants to use; the priority order set by the hub does not have to be observed.

Not really.

There is **no** DC client with a built-in cipher detection mechanism except the newest DC++ 0.870. That means, for the vast majority of users the first cipher offered will be used.

> ciphers = "HIGH+kEDH:HIGH+kEECDH:HIGH:!PSK:!SRP:!3DES:!aNULL"

Btw still do not understand what is that string for. There are no 3DES ciphers for TLS 1.3, no PSK or NULLs, no insecure ciphers at all...

**Comment by pulsar-de** [comment]
The order of the ciphers used by the client is decisive. e.g.: AirDC++ uses in his [CryptoManager](https://github.com/airdcpp/airdcpp-windows/blob/master/airdcpp/airdcpp/CryptoManager.cpp#L145) this order:
```cpp
	const char ciphersuitesTls13[] =
		"TLS_AES_128_GCM_SHA256:"
		"TLS_AES_256_GCM_SHA384:"
		"TLS_CHACHA20_POLY1305_SHA256";
```
As you can see, the "TLS_AES_128_GCM_SHA256" is the first cipher in the AirDC priority order. if the hub offers this, it will also be used. AirDC only jumps to the next cipher in the list if the hub does not offer the first. If the hub offers these ciphers:
```lua
ciphersuites = "TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256",
```
So AirDC will using "TLS_AES_128_GCM_SHA256" because of his priority order.

The hub therefore offers a selection of available ciphers and the client then decides which one to take based on his priority list. in order to force a cipher you have to limit the selection to one cipher. 

> Btw still do not understand what is that string for. There are no 3DES ciphers for TLS 1.3, no PSK or NULLs, no insecure ciphers at all...

Of course you are absolutely right, that is unnecessary. I just hadn't changed that yet. it is completely sufficient if, for example, "HIGH" is defined: 
```lua
ciphers = "HIGH",
```

**Comment by Tantrix** [comment]
Uhu. Did some testing. Seems like you are right about the order of the ciphers.

But.

> The order of the ciphers used by the client is decisive
> the client decides which cipher he wants to use

From the hub's (and hub admin too!) point of view that is not the same. 

I.e. if I **do not** want **TLS_AES_256_GCM_SHA384** cipher to be used I will disable it totally, independently from DC client _wants_ to use. And it will comply.

That means, **TLS_AES_128_GCM_SHA256** or **ChaCha** cipher will be used. What exactly? Depends on the order of the ciphers set by the client dev (excepting DC++ 0.870+), as you mentioned above.

So I can't call a client's choice _decisive_.

In other words, there are exactly two options for me as for hub admin:

1) Do not set TLS 1.3 ciphers list at all, leave it default.
2) Limit the TLS 1.3 cipher's choice.

That's all I wanted to say :)

**Comment by Tantrix** [comment]
Thanks.

`				ciphersuites = "TLS_CHACHA20_POLY1305_SHA256",`

works for me.

2.22
